File: packages/native-android/android/src/main/java/com/openota/runtime/BundleVerifier.kt
package com.openota.runtime

import java.io.File
import java.security.MessageDigest

private const val ANDROID_PLATFORM = "android"
private const val HASH_BUFFER_SIZE = 1 shl 16 // 64 KiB

/** The outcome of a full verification pass: the parsed manifest plus its already-computed hash. */
data class VerifiedBundle(
    val packageDir: File,
    val manifest: BundleManifest,
    val bundleFile: File,
    val sha256: String,
)

/**
 * Re-verifies a staged package independently of whatever checks the JS SDK already performed.
 * The JS layer is not a trust boundary — it can be bypassed, patched, or simply buggy — so every
 * field the runtime depends on to decide "is this bundle safe to serve" is re-derived from the
 * bytes on disk here, once, and the result ([VerifiedBundle]) is threaded through the rest of the
 * pipeline so [BundleInstaller] never has to hash the file again.
 */
class BundleVerifier(private val storage: BundleStorage, private val appRuntimeVersion: String) {

    fun verify(packageDir: File): VerifiedBundle {
        storage.assertNoSymlinks(packageDir)

        val manifestFile = storage.manifestFile(packageDir)
        if (!manifestFile.exists()) {
            throw BundleVerificationException("Manifest not found at \"${manifestFile.path}\"")
        }

        val manifest = BundleManifest.parse(manifestFile.readText())

        if (manifest.platform != ANDROID_PLATFORM) {
            throw BundleVerificationException("Manifest platform \"${manifest.platform}\" is not \"$ANDROID_PLATFORM\"")
        }

        if (manifest.runtimeVersion != appRuntimeVersion) {
            throw UnsupportedRuntimeVersionException(manifest.runtimeVersion, appRuntimeVersion)
        }

        val bundleFile = storage.bundleFile(packageDir, manifest.bundleName)
        if (!bundleFile.exists() || !bundleFile.isFile) {
            throw BundleVerificationException("Bundle file not found at \"${bundleFile.path}\"")
        }

        // Packages with no static assets may not have an assets/ directory at all: the CLI omits
        // it, and even when present as an empty zip entry, some Android unzip implementations
        // don't materialize empty directories on extraction. Only a *non-directory* in its place
        // is an error; a missing directory simply means "no assets".
        val assetsDir = storage.assetsDir(packageDir)
        if (assetsDir.exists() && !assetsDir.isDirectory) {
            throw BundleVerificationException("Assets path is not a directory at \"${assetsDir.path}\"")
        }

        val actualSha256 = OpenOTALogger.timed("verify.sha256") { computeSha256(bundleFile) }

        if (!actualSha256.equals(manifest.sha256, ignoreCase = true)) {
            throw BundleVerificationException(
                "Checksum mismatch for ${manifest.platform}@${manifest.version}: " +
                    "expected ${manifest.sha256}, got $actualSha256",
            )
        }

        return VerifiedBundle(packageDir, manifest, bundleFile, actualSha256)
    }

    /** Streams the file through a digest in fixed-size chunks; the whole file is never buffered in memory. */
    private fun computeSha256(file: File): String {
        val digest = MessageDigest.getInstance("SHA-256")
        val buffer = ByteArray(HASH_BUFFER_SIZE)

        file.inputStream().buffered().use { input ->
            while (true) {
                val read = input.read(buffer)
                if (read < 0) break
                digest.update(buffer, 0, read)
            }
        }

        return digest.digest().joinToString(separator = "") { "%02x".format(it) }
    }
}

File: packages/native-ios/ios/BundleVerifier.swift
import Foundation
import CryptoKit

private let iosPlatform = "ios"

/// The outcome of a full verification pass: the parsed manifest plus its already-computed hash.
struct VerifiedBundle {
    let packageDir: URL
    let manifest: BundleManifest
    let bundleFile: URL
    let sha256: String
}

/// Re-verifies a staged package independently of whatever checks the JS SDK already performed —
/// mirrors `BundleVerifier.kt`. JS is not a trust boundary, so every field the runtime depends on
/// is re-derived from the bytes on disk here, once.
final class BundleVerifier {
    private let storage: BundleStorage
    private let appRuntimeVersion: String

    init(storage: BundleStorage, appRuntimeVersion: String) {
        self.storage = storage
        self.appRuntimeVersion = appRuntimeVersion
    }

    func verify(_ packageDir: URL) throws -> VerifiedBundle {
        try storage.assertNoSymlinks(packageDir)

        let manifestFile = storage.manifestFile(packageDir)
        guard storage.fileExists(manifestFile) else {
            throw OpenOTAException.verificationFailed("Manifest not found at \"\(manifestFile.path)\"")
        }
        let manifest = try BundleManifest.parse(contentsOf: manifestFile)

        guard manifest.platform == iosPlatform else {
            throw OpenOTAException.verificationFailed("Manifest platform \"\(manifest.platform)\" is not \"\(iosPlatform)\"")
        }

        guard manifest.runtimeVersion == appRuntimeVersion else {
            throw OpenOTAException.invalidRuntime(bundleRuntimeVersion: manifest.runtimeVersion, appRuntimeVersion: appRuntimeVersion)
        }

        let bundleFile = storage.bundleFile(packageDir, bundleFileName: manifest.bundleName)
        guard storage.fileExists(bundleFile, isDirectory: false) else {
            throw OpenOTAException.verificationFailed("Bundle file not found at \"\(bundleFile.path)\"")
        }

        // Packages with no static assets may not ship an assets/ directory at all. Only a
        // *non-directory* in its place is an error; missing simply means "no assets".
        let assetsDir = storage.assetsDir(packageDir)
        if storage.fileExists(assetsDir), !storage.fileExists(assetsDir, isDirectory: true) {
            throw OpenOTAException.verificationFailed("Assets path is not a directory at \"\(assetsDir.path)\"")
        }

        let actualSha256 = try OpenOTALogger.timed("verify.sha256") { try Self.computeSha256(bundleFile) }
        guard actualSha256.caseInsensitiveCompare(manifest.sha256) == .orderedSame else {
            throw OpenOTAException.verificationFailed(
                "Checksum mismatch for \(manifest.platform)@\(manifest.version): expected \(manifest.sha256), got \(actualSha256)"
            )
        }

        return VerifiedBundle(packageDir: packageDir, manifest: manifest, bundleFile: bundleFile, sha256: actualSha256)
    }

    /// Streams the file through SHA-256 in fixed-size chunks so the whole file is never buffered
    /// in memory, mirroring the Kotlin implementation's `MessageDigest` streaming approach.
    static func computeSha256(_ url: URL) throws -> String {
        guard let stream = InputStream(url: url) else {
            throw OpenOTAException.verificationFailed("Bundle file is unreadable at \"\(url.path)\"")
        }
        stream.open()
        defer { stream.close() }

        var hasher = SHA256()
        let bufferSize = 1 << 16 // 64 KiB
        var buffer = [UInt8](repeating: 0, count: bufferSize)

        while stream.hasBytesAvailable {
            let read = stream.read(&buffer, maxLength: bufferSize)
            if read < 0 {
                throw OpenOTAException.verificationFailed("Failed reading bundle file at \"\(url.path)\"")
            }
            if read == 0 { break }
            hasher.update(data: Data(buffer[0..<read]))
        }

        let digest = hasher.finalize()
        return digest.map { String(format: "%02x", $0) }.joined()
    }
}
